### PR TITLE
Log fewer errors

### PR DESCRIPTION
**Pull request by sambitesh**
This is a continuation of commit 456b2b3 in GPORCA. Adding more errors to the list that
doesn't get logged in log file.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/gporca/libgpos/include/gpos/_api.h [hunk 1]
```diff
@@ -29,7 +29,7 @@ gpos::BOOL FoundException(gpos::CException &exc, const gpos::ULONG *exceptions,
 
 // Check if given exception is an unexpected reason for failing to
 // produce a plan
-gpos::BOOL IsUnexpectedFailure(gpos::CException &exc);
+gpos::BOOL IsLoggableFailure(gpos::CException &exc);
```
Comment (chrishajas) [inline src/backend/gporca/libgpos/include/gpos/_api.h:32]: How do we decide whether we should log an error? Are we still differentiating between expected vs unexpected fallbacks?

File: src/backend/gporca/libgpos/src/_api.cpp [hunk 2]
```diff
@@ -30,22 +30,42 @@
 
 using namespace gpos;
 
+
+// refer gpopt/exception.cpp for explanation of errors
 const ULONG expected_opt_fallback[] =
```
Comment (chrishajas) [inline src/backend/gporca/libgpos/src/_api.cpp:35]: Are these still expected fallbacks?

**Comment by zellerh** [comment]
Wouldn't it be nice if we could control the level of logging with the `log_min_messages` guc? I haven't studied this in detail, but my guess is that when Orca throws an exception, it doesn't specify a severity level (it leaves the `CException::m_severity_level` field unset), and we log the exception at the log level `error`. One possible fix might be to set the severity (log level) of some exceptions to `warning`, `notice` or `info`, so they could be suppressed with the `log_min_messages` guc set to `error` or higher. See https://community.pivotal.io/s/article/How-to-Limit-Pivotal-Greenplum-Logging-Information?language=en_US.

**Comment by zellerh** [comment]
Sorry, I didn't understand when I made my earlier comment that the issue is really about extra messages on top of the messages that describe the root cause. As far as I can tell, we are talking about the following two:

```
LOG: "Pivotal Optimizer (GPORCA) failed to produce plan"
```

This is produced based on the `optimizer_log_failure` guc (0 means log all fallbacks, 1 only unexpected ones (the default) and 2 means log only expected failures).

```
LOG: "3rd party error log: Unexpected exception reached top of execution stack"
```

This second message is generated only for unexpected failures. It comes out in a somewhat strange way and as two log events, since it is produced by writing to std::cerr instead of calling using the elog macro in GPDB.

So, maybe we could combine these two into one:

```
LOG: "Pivotal Optimizer (GPORCA) failed to produce plan (unexpected)"
```

**Comment by chrishajas** [comment]
I agree that the `Unexpected exception reached top of execution stack` is repetitive and should be removed. I'm not sure if we want to add all of these exceptions as expected though in the first commit, as some of them don't seem expected (eg: `ExmiDXLIncorrectNumberOfChildren` doesn't seem expected).

**Comment by sambitesh** [comment]
@chrishajas my argument was only those errors which are due to the system like out of mem should make it to the log file. Other errors which can easily be reproed locally shouldn't. We can always check those by setting `client_min_messages to log`

**Comment by zellerh** [comment]
With the latest commit it doesn't matter a great deal whether an error is considered "expected" or "loggable".

Here is an example. The first few lines is what gets logged currently for an unexpected error (4 messages):

```
2020-05-18 17:20:42.768908 PDT,,,p63499,th363267520,,,,0,,,seg-1,,,,,"LOG","00000","3rd party error log:
Unexpected exception reached top of execution stack: major=1",,,,,,,,"SysLoggerMain","syslogger.c",638,
2020-05-18 17:20:42.768995 PDT,,,p63499,th363267520,,,,0,,,seg-1,,,,,"LOG","00000","3rd party error log:
 minor=2 file=../libgpopt/src/operators/CExpressionHandle.cpp line=1243",,,,,,,,"SysLoggerMain","syslogger.c",653,
2020-05-18 17:20:42.769097 PDT,"hzeller","hzeller",p63697,th363267520,"::1","52666",2020-05-18 17:19:54 PDT,0,con14,cmd3,seg-1,,dx3,,sx1,"LOG","00000","2020-05-18 17:20:42:767948 PDT,THD000,ERROR,""../libgpopt/src/operators/CExpressionHandle.cpp:1243: Failed assertion: __null == m_pgexpr"",",,,,,,"explain select count(distinct a), count(distinct b) from foo;",0,,"COptTasks.cpp",566,
2020-05-18 17:20:42.769172 PDT,"hzeller","hzeller",p63697,th363267520,"::1","52666",2020-05-18 17:19:54 PDT,0,con14,cmd3,seg-1,,dx3,,sx1,"LOG","00000","Pivotal Optimizer (GPORCA) failed to produce plan",,,,,,"explain select count(distinct a), count(distinct b) from foo;",0,,"orca.c",62,
```
Here is a simulation of what would get logged with this PR:

```
2020-05-18 17:20:42.769097 PDT,"hzeller","hzeller",p63697,th363267520,"::1","52666",2020-05-18 17:19:54 PDT,0,con14,cmd3,seg-1,,dx3,,sx1,"LOG","00000","2020-05-18 17:20:42:767948 PDT,THD000,ERROR,""../libgpopt/src/operators/CExpressionHandle.cpp:1243: Failed assertion: __null == m_pgexpr"",",,,,,,"explain select count(distinct a), count(distinct b) from foo;",0,,"COptTasks.cpp",566,
2020-05-18 17:20:42.769172 PDT,"hzeller","hzeller",p63697,th363267520,"::1","52666",2020-05-18 17:19:54 PDT,0,con14,cmd3,seg-1,,dx3,,sx1,"LOG","00000","Pivotal Optimizer (GPORCA) failed to produce plan",,,,,,"explain select count(distinct a), count(distinct b) from foo;",0,,"orca.c",62,
```

The only difference between loggable and not loggable errors is now that the former have
"(unexpected)" in the second message. The version shown above is now an expected error, it does not have the word in it.

**Comment by chrishajas** [comment]
If the only difference between expected/unexpected errors is the `(unexpected)` text in `LOG: "Pivotal Optimizer (GPORCA) failed to produce plan (unexpected)`, should we get rid of the distinction logic? I could see it being useful if we encouraged users to report unexpected issues as bugs, but I don't think we plan to do that. Currently there isn't a different action they should take if it is expected vs unexpected.

**Comment by zellerh** [comment]
Don't you think that it might be useful for users' own analysis? I would assume users won't report unexpected fallbacks right away, but maybe they might investigate a bit. For example, this might be useful in debugging an application after upgrading to a new version.

**Comment by chrishajas** [comment]
> Don't you think that it might be useful for users' own analysis? I would assume users won't report unexpected fallbacks right away, but maybe they might investigate a bit. For example, this might be useful in debugging an application after upgrading to a new version.

Sounds good, this looks good to me then. I keep forgetting what rules we're using to determine if something is "unexpected", which I think as of this PR is anything not user-configuration (out of mem, etc.) related.

**Pull request merged**